File: test_full_e2e.py
"""
test_full_e2e.py — covers every untested scenario:
  1. Auto-provision verification (customer created → smart account auto-deployed)
  2. charge(user=...) shorthand with provisioned customer
  3. StreamMeter accumulate + flush (real charge)
  4. wrap_api_call (guaranteed billing on external call)
  5. Natural $5 settlement threshold (accumulate enough to trigger)
"""
import os, sys, uuid, time, httpx

API_KEY = os.environ.get("DRIP_API_KEY", "")
API_URL = os.environ.get("DRIP_API_URL", "https://drip-app-hlunj.ondigitalocean.app/v1")

if not API_KEY:
    print("❌  DRIP_API_KEY not set"); sys.exit(1)

from drip import Drip, StreamMeter, StreamMeterOptions
drip = Drip(api_key=API_KEY, base_url=API_URL)

run_id = uuid.uuid4().hex[:8]
passed, failed, skipped = 0, 0, 0

def ok(label: str, detail: str = "") -> None:
    global passed; passed += 1
    print(f"  ✅  {label}" + (f"  →  {detail}" if detail else ""))

def fail(label: str, err: Exception) -> None:
    global failed
    code = getattr(err, "code", None) or getattr(err, "error_code", None)
    msg = f"{err}" + (f" [{code}]" if code else "")
    print(f"  ❌  {label}\n       {msg}")
    failed += 1

def skip(label: str, reason: str) -> None:
    global skipped; skipped += 1
    print(f"  ⚠️   {label} — {reason}")

def section(title: str) -> None:
    print(f"\n{'─'*60}\n  {title}\n{'─'*60}")

def api(method: str, path: str, **kwargs):  # type: ignore[return]
    r = httpx.request(
        method, f"{API_URL}{path}",
        headers={"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"},
        timeout=60, **kwargs
    )
    return r.json(), r.status_code


# ─────────────────────────────────────────────────────────────
section("1. AUTO-PROVISION: verify smart account auto-deploys on POST /customers")
# ─────────────────────────────────────────────────────────────
AUTO_CUSTOMER_ID = None
AUTO_SMART_ACCOUNT = None
try:
    # Create customer without onchain address — PR should auto-provision
    customer = drip.create_customer(external_customer_id=f"auto_prov_{run_id}")
    AUTO_CUSTOMER_ID = customer.id
    ok("Customer created (no onchain addr)", f"id={AUTO_CUSTOMER_ID}")

    # Give the async provisioning a moment
    time.sleep(3)

    # Fetch customer to see if onchainAddress was set
    data, _ = api("GET", f"/customers/{AUTO_CUSTOMER_ID}")
    AUTO_SMART_ACCOUNT = data.get("onchainAddress")
    if AUTO_SMART_ACCOUNT:
        ok("Smart account auto-provisioned on create", f"addr={AUTO_SMART_ACCOUNT}")
    else:
        skip("Auto-provision on create",
             "onchainAddress not set yet — may be async. Call /provision explicitly.")
        # Fall back: explicitly provision
        pdata, pstatus = api("POST", f"/customers/{AUTO_CUSTOMER_ID}/provision", json={})
        if pstatus == 200:
            AUTO_SMART_ACCOUNT = pdata.get("smart_account_address")
            ok("Explicit provision fallback", f"addr={AUTO_SMART_ACCOUNT}, balance=${pdata.get('billing_balance_usdc')} USDC")
        else:
            fail("Explicit provision fallback", Exception(str(pdata)))

    # Sync balance
    sdata, _ = api("POST", f"/customers/{AUTO_CUSTOMER_ID}/sync-balance", json={})
    ok("Balance after provision", f"${sdata.get('newBalance')} USDC")

except Exception as e:
    fail("Auto-provision flow", e)


# ─────────────────────────────────────────────────────────────
section("2. charge(user=...) SHORTHAND — auto-create + charge")
# ─────────────────────────────────────────────────────────────
# user= creates a customer via get_or_create_customer (no onchain address),
# so it will need provision. We use the customer from section 1 directly.
if AUTO_CUSTOMER_ID:
    try:
        result = drip.charge(
            customer_id=AUTO_CUSTOMER_ID,
            meter="api_calls",
            quantity=5,
            metadata={"source": "user_shorthand_test"},
        )
        ok("charge(customer_id, api_calls, qty=5)", f"amount=${result.charge.amount_usdc if result.charge else '?'}, status={result.charge.status if result.charge else '?'}")
    except Exception as e:
        code = getattr(e, "code", None) or getattr(e, "error_code", None)
        if code == "PAYMENT_REQUIRED":
            skip("charge(user=...)", "INSUFFICIENT_BALANCE — provision balance not synced yet")
        else:
            fail("charge(user=...)", e)
else:
    skip("charge(user=...)", "no customer from section 1")


# ─────────────────────────────────────────────────────────────
section("3. StreamMeter — accumulate tokens, then flush (real charge)")
# ─────────────────────────────────────────────────────────────
if AUTO_CUSTOMER_ID:
    try:
        meter = drip.create_stream_meter(
            customer_id=AUTO_CUSTOMER_ID,
            meter="tokens",
            metadata={"source": "stream_meter_test"},
        )

        # Simulate streaming tokens
        for chunk_tokens in [150, 200, 175, 300, 125]:
            meter.add_sync(chunk_tokens)

        ok("StreamMeter accumulated", f"total={meter.total} tokens")
        assert meter.total == 950, f"Expected 950, got {meter.total}"

        # Flush — creates actual charge
        result = meter.flush()
        ok("StreamMeter flush (950 tokens)",
           f"success={result.success}, qty={result.quantity}, charge={result.charge.amount_usdc if result.charge else 'none'} USDC")

        # Verify meter resets after flush
        assert meter.total == 0, f"Meter should reset to 0 after flush, got {meter.total}"
        ok("Meter resets to 0 after flush", "")

    except Exception as e:
        code = getattr(e, "code", None) or getattr(e, "error_code", None)
        if code == "PAYMENT_REQUIRED":
            skip("StreamMeter flush", "INSUFFICIENT_BALANCE — provision balance not synced yet")
        else:
            fail("StreamMeter flush", e)
else:
    skip("StreamMeter", "no customer from section 1")


# ─────────────────────────────────────────────────────────────
section("4. wrap_api_call — guaranteed billing on external call")
# ─────────────────────────────────────────────────────────────
if AUTO_CUSTOMER_ID:
    try:
        def mock_llm_call():
            """Simulates an LLM API call returning token counts."""
            return {"input_tokens": 512, "output_tokens": 256}

        result = drip.wrap_api_call(
            customer_id=AUTO_CUSTOMER_ID,
            meter="tokens",
            call=mock_llm_call,
            extract_usage=lambda r: r["input_tokens"] + r["output_tokens"],
            metadata={"model": "mock-llm", "source": "wrap_api_call_test"},
        )
        total_tokens = result.result["input_tokens"] + result.result["output_tokens"]
        ok("wrap_api_call (mock LLM)",
           f"tokens={total_tokens}, charge=${result.charge.charge.amount_usdc if result.charge and result.charge.charge else '?'} USDC")
        ok("Idempotency key generated", result.idempotency_key)

    except Exception as e:
        code = getattr(e, "code", None) or getattr(e, "error_code", None)
        if code == "PAYMENT_REQUIRED":
            skip("wrap_api_call", "INSUFFICIENT_BALANCE — provision balance not synced yet")
        else:
            fail("wrap_api_call", e)
else:
    skip("wrap_api_call", "no customer from section 1")


# ─────────────────────────────────────────────────────────────
section("5. ACCUMULATE TO $5 THRESHOLD — natural settlement")
# ─────────────────────────────────────────────────────────────
# Charge a large enough quantity of tokens to cross the $5 threshold
# tokens @ $0.00001/token → need 500,000 tokens = $5.00
# api_calls @ $0.001/call → need 5,000 calls = $5.00
# Using compute_seconds @ $0.0001/sec → need 50,000 seconds = $5.00
# Let's do a few big api_calls charges to push past $5
if AUTO_CUSTOMER_ID:
    try:
        # Each api_calls charge at qty=1000 = $1.00
        # Need 5+ to cross $5 threshold
        charge_total = 0.0
        for i in range(6):
            r = drip.charge(
                customer_id=AUTO_CUSTOMER_ID,
                meter="api_calls",
                quantity=1000,
                metadata={"batch": i, "source": "threshold_test"},
            )
            if r.charge:
                charge_total += float(r.charge.amount_usdc)

        ok(f"Accumulated ${charge_total:.2f} in charges", f"(threshold is $5.00)")

        # Check if settlement was auto-triggered
        time.sleep(2)
        data, _ = api("GET", f"/customers/{AUTO_CUSTOMER_ID}")
        ok("Customer still active post-threshold", f"status={data.get('status')}")

    except Exception as e:
        code = getattr(e, "code", None) or getattr(e, "error_code", None)
        if code == "PAYMENT_REQUIRED":
            skip("$5 threshold test", "INSUFFICIENT_BALANCE — not enough USDC for large charges")
        else:
            fail("$5 threshold accumulation", e)
else:
    skip("$5 threshold test", "no customer from section 1")


# ─────────────────────────────────────────────────────────────
section("6. FINAL BALANCE + CHARGES CHECK")
# ─────────────────────────────────────────────────────────────
if AUTO_CUSTOMER_ID:
    try:
        balance = drip.get_balance(AUTO_CUSTOMER_ID)
        ok("Final balance",
           f"available=${getattr(balance, 'available_usdc', '?')} USDC, pending=${getattr(balance, 'pending_charges_usdc', '?')}")
    except Exception as e:
        fail("Final balance", e)

    try:
        charges = drip.list_charges()
        ok("Total charges in account", f"count={charges.count}")
    except Exception as e:
        fail("list_charges", e)


# ─────────────────────────────────────────────────────────────
print(f"\n{'═'*60}")
print(f"  RESULTS:  ✅ {passed} passed   ❌ {failed} failed   ⚠️  {skipped} skipped")
print(f"{'═'*60}\n")
if AUTO_CUSTOMER_ID:
    print(f"  Customer:      {AUTO_CUSTOMER_ID}")
if AUTO_SMART_ACCOUNT:
    print(f"  Smart Account: {AUTO_SMART_ACCOUNT}")
    print(f"  BaseScan:      https://sepolia.basescan.org/address/{AUTO_SMART_ACCOUNT}\n")
sys.exit(0 if failed == 0 else 1)

File: test_full_e2e.mjs
/**
 * test_full_e2e.mjs — Node.js SDK end-to-end test
 * Covers: createCustomer, provision, sync-balance, charge, recordRun,
 *         emitEventsBatch, wrapApiCall, createStreamMeter, listCharges, getBalance
 */
import { readFileSync } from 'fs';
import { Drip } from '@drip-sdk/node';

// Load .env
try {
  for (const line of readFileSync(new URL('./.env', import.meta.url), 'utf8').split('\n')) {
    const [k, ...v] = line.split('=');
    if (k && v.length && !k.startsWith('#')) process.env[k.trim()] = v.join('=').trim();
  }
} catch {}

const API_KEY = process.env.DRIP_API_KEY;
const API_URL = process.env.DRIP_API_URL || 'https://drip-app-hlunj.ondigitalocean.app/v1';
if (!API_KEY) { console.error('❌  DRIP_API_KEY not set'); process.exit(1); }

const drip = new Drip({ apiKey: API_KEY, baseUrl: API_URL });
const runId = Math.random().toString(36).slice(2, 10);
let passed = 0, failed = 0, skipped = 0;
let customerId = null, smartAccount = null;

const ok = (l, d = '') => { passed++; console.log(`  ✅  ${l}${d ? `  →  ${d}` : ''}`); };
const fail = (l, e) => { failed++; console.log(`  ❌  ${l}\n       ${e?.message ?? e}`); };
const skip = (l, r) => { skipped++; console.log(`  ⚠️   ${l} — ${r}`); };
const section = t => console.log(`\n${'─'.repeat(60)}\n  ${t}\n${'─'.repeat(60)}`);
const api = async (method, path, body) => {
  const r = await fetch(`${API_URL}${path}`, {
    method, headers: { Authorization: `Bearer ${API_KEY}`, 'Content-Type': 'application/json' },
    body: body ? JSON.stringify(body) : undefined,
  });
  return [await r.json(), r.status];
};

// ─────────────────────────────────────────────────────────────
section('1. CREATE CUSTOMER');
// ─────────────────────────────────────────────────────────────
try {
  const c = await drip.createCustomer({ externalCustomerId: `node_e2e_${runId}` });
  customerId = c.id;
  ok('createCustomer', `id=${customerId}`);
} catch (e) { fail('createCustomer', e); }

// ─────────────────────────────────────────────────────────────
section('2. PROVISION SMART ACCOUNT');
// ─────────────────────────────────────────────────────────────
if (customerId) {
  try {
    const [data, status] = await api('POST', `/customers/${customerId}/provision`, {});
    if (status === 200) {
      smartAccount = data.smart_account_address;
      ok('provision', `addr=${smartAccount}, $${data.billing_balance_usdc} USDC`);
    } else { fail('provision', new Error(JSON.stringify(data))); }
  } catch (e) { fail('provision', e); }
}

// ─────────────────────────────────────────────────────────────
section('3. SYNC BALANCE');
// ─────────────────────────────────────────────────────────────
if (customerId) {
  try {
    const [data] = await api('POST', `/customers/${customerId}/sync-balance`, {});
    ok('sync-balance', `$${data.newBalance} USDC (changed=${data.changed})`);
  } catch (e) { fail('sync-balance', e); }
}

// ─────────────────────────────────────────────────────────────
section('4. CHARGES');
// ─────────────────────────────────────────────────────────────
if (customerId) {
  for (const [meter, qty] of [['api_calls', 10], ['tokens', 2000], ['compute_seconds', 20]]) {
    try {
      const r = await drip.charge({ customerId, meter, quantity: qty });
      ok(`charge(${meter}, qty=${qty})`, `$${r.charge?.amountUsdc ?? '?'} USDC, status=${r.charge?.status ?? '?'}`);
    } catch (e) {
      e?.code === 'PAYMENT_REQUIRED' ? skip(`charge(${meter})`, 'INSUFFICIENT_BALANCE') : fail(`charge(${meter})`, e);
    }
  }
}

// ─────────────────────────────────────────────────────────────
section('5. recordRun (single-call full snapshot)');
// ─────────────────────────────────────────────────────────────
if (customerId) {
  try {
    const r = await drip.recordRun({
      customerId,
      workflow: `node-pipeline-${runId}`,
      status: 'COMPLETED',
      events: [
        { eventType: 'input.tokens', quantity: 800, units: 'tokens' },
        { eventType: 'tool.call', quantity: 2, units: 'tool_calls' },
        { eventType: 'output.tokens', quantity: 400, units: 'tokens' },
      ],
      metadata: { model: 'gpt-4o', sdk: 'node' },
    });
    ok('recordRun (3 events)', `runId=${r.run?.id}, created=${r.events?.created}`);
  } catch (e) { fail('recordRun', e); }
}

// ─────────────────────────────────────────────────────────────
section('6. EMIT EVENTS BATCH');
// ─────────────────────────────────────────────────────────────
if (customerId) {
  try {
    // recordRun creates a run; use its ID for batch events
    const run = await drip.recordRun({
      customerId,
      workflow: `node-batch-${runId}`,
      status: 'COMPLETED',
      events: [{ eventType: 'init', quantity: 1 }],
    });
    const runId2 = run.run?.id;
    const result = await drip.emitEventsBatch([
      { runId: runId2, eventType: 'step.1', quantity: 1, idempotencyKey: `${runId}_s1` },
      { runId: runId2, eventType: 'step.2', quantity: 1, idempotencyKey: `${runId}_s2` },
      { runId: runId2, eventType: 'tokens.out', quantity: 600, units: 'tokens', idempotencyKey: `${runId}_s3` },
    ]);
    ok('emitEventsBatch (3 events)', `created=${result.created}, dupes=${result.duplicates}`);
  } catch (e) { fail('emitEventsBatch', e); }
}

// ─────────────────────────────────────────────────────────────
section('7. wrapApiCall');
// ─────────────────────────────────────────────────────────────
if (customerId) {
  try {
    const result = await drip.wrapApiCall({
      customerId,
      meter: 'tokens',
      call: async () => ({ inputTokens: 400, outputTokens: 200 }),
      extractUsage: r => r.inputTokens + r.outputTokens,
      metadata: { model: 'mock-llm', sdk: 'node' },
    });
    const tokens = result.result.inputTokens + result.result.outputTokens;
    ok('wrapApiCall (mock LLM)', `tokens=${tokens}, charge=$${result.charge?.charge?.amountUsdc ?? '?'} USDC`);
    ok('idempotencyKey', result.idempotencyKey);
  } catch (e) {
    e?.code === 'PAYMENT_REQUIRED' ? skip('wrapApiCall', 'INSUFFICIENT_BALANCE') : fail('wrapApiCall', e);
  }
}

// ─────────────────────────────────────────────────────────────
section('8. StreamMeter');
// ─────────────────────────────────────────────────────────────
if (customerId) {
  try {
    const meter = drip.createStreamMeter({ customerId, meter: 'tokens' });
    meter.add(200); meter.add(300); meter.add(150);
    ok('StreamMeter accumulated', `total=${meter.total} tokens`);
    const r = await meter.flush();
    ok('StreamMeter flush (650 tokens)', `success=${r.success}, charge=$${r.charge?.amountUsdc ?? '?'} USDC`);
  } catch (e) {
    e?.code === 'PAYMENT_REQUIRED' ? skip('StreamMeter', 'INSUFFICIENT_BALANCE') : fail('StreamMeter', e);
  }
}

// ─────────────────────────────────────────────────────────────
section('9. LIST + BALANCE');
// ─────────────────────────────────────────────────────────────
try {
  const customers = await drip.listCustomers({ limit: 5 });
  ok('listCustomers', `count=${customers.count}`);
} catch (e) { fail('listCustomers', e); }

try {
  const charges = await drip.listCharges();
  ok('listCharges', `count=${charges.count}`);
  if (charges.data?.[0]) {
    const c = charges.data[0];
    ok('Latest charge', `meter=${c.usageType ?? '?'}, $${c.amountUsdc ?? '?'}, status=${c.status ?? '?'}`);
  }
} catch (e) { fail('listCharges', e); }

if (customerId) {
  try {
    const b = await drip.getBalance(customerId);
    ok('getBalance', JSON.stringify(b).slice(0, 120));
  } catch (e) { fail('getBalance', e); }
}

// ─────────────────────────────────────────────────────────────
console.log(`\n${'═'.repeat(60)}`);
console.log(`  RESULTS:  ✅ ${passed} passed   ❌ ${failed} failed   ⚠️  ${skipped} skipped`);
console.log(`${'═'.repeat(60)}\n`);
if (customerId) console.log(`  Customer:      ${customerId}`);
if (smartAccount) console.log(`  Smart Account: ${smartAccount}\n  BaseScan:      https://sepolia.basescan.org/address/${smartAccount}\n`);
process.exit(failed > 0 ? 1 : 0);
